interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+// this function directly.
+//
+// Constructs a new rule with the given parameters as values, with the
+// exception of duration. Uses the given duration, in addition to the current
+// time, to compute the expiration time for the rule, and stores that as part
+// of the rule which is returned. If any of the given parameters are invalid,
+// returns a corresponding error.
+//
+// TODO: unexport (probably, avoid confusion with AddRule)
+func (rdb *RuleDB) PopulateNewRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (*Rule, error) {
+	if err := constraints.ValidateForInterface(iface); err != nil {
+		return nil, err
+	}
+	if _, err := outcome.AsBool(); err != nil {
+		// This should not occur, since PopulateNewRule should only be called
+		// on values which were validated while unmarshalling
+		return nil, err
+	}
+	id, _ := rdb.maxIDMmap.NextID()
+	currTime := time.Now()
+	expiration, err := lifespan.ParseDuration(duration, currTime)
+	if err != nil {
+		return nil, err
+	}
+	newRule := Rule{
+		ID:          id,
+		Timestamp:   currTime,
+		User:        user,
+		Snap:        snap,
+		Interface:   iface,
+		Constraints: constraints,
+		Outcome:     outcome,
+		Lifespan:    lifespan,
+		Expiration:  expiration,
+	}
+	return &newRule, nil
+}
+
+// IsPathAllowed checks whether the given path with the given permission is
+// allowed or denied by existing rules for the given user, snap, and interface.
+// If no rule applies, returns ErrNoMatchingRule.
+func (rdb *RuleDB) IsPathAllowed(user uint32, snap string, iface string, path string, permission string) (bool, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	needToSave := false
+	defer func() {
+		if needToSave {
+			rdb.save()
+		}
+	}()
+	variantMap := rdb.permissionDBForUserSnapInterfacePermission(user, snap, iface, permission).VariantEntries
+	matchingVariants := make([]patterns.PatternVariant, 0)
+	// Make sure all rules use the same expiration timestamp, so a rule with
+	// an earlier expiration cannot outlive another rule with a later one.
+	currTime := time.Now()
+	for variantStr, variantEntry := range variantMap {
+		matchingRule, err := rdb.ruleWithID(variantEntry.RuleID)
+		if err != nil {
+			// Database was left inconsistent, should not occur
+			delete(variantMap, variantStr)
+			// Record a notice for the offending rule, just in case
+			rdb.notifyRule(user, variantEntry.RuleID, nil)
+			continue
+		}
+		expired, err := matchingRule.Expired(currTime)
+		switch {
+		case err != nil:
+			// Should not occur
+			logger.Noticef("error while checking whether rule had expired: %v", err)
+			fallthrough
+		case expired:
+			needToSave = true
+			rdb.removeRuleFromTree(matchingRule)
+			rdb.removeRuleWithID(variantEntry.RuleID)
+			data := map[string]string{"removed": "expired"}
+			rdb.notifyRule(user, variantEntry.RuleID, data)
```

Comment:
do we need to do this on the fly?

Reply:
No, now we only expire rules on `load()` or when there's a conflict with a new rule.